File: r/tests/testthat/test_edge_cases.R
context("Test edge cases")

test_that("Node size is respected", {
  n <- 1000
  x <- matrix(runif(n))
  z <- matrix(runif(n))

  n_trees <- 1
  mtry <- 1
  n_basis <- 15
  for(min_size in c(1, 2, 3, 10, 100)) {
    forest <- RFCDE(x, z, n_trees, mtry, min_size, n_basis)
    expect_gte(sum(weights(forest, x[1, ])), min_size)
  }
})

test_that("Binary splits are respected", {
  n <- 1000
  x <- matrix(sample(1:2, n, replace = TRUE))
  z <- matrix(rnorm(n))

  n_trees <- 1
  mtry <- 1
  min_size <- 20
  n_basis <- 15
  forest <- RFCDE(x, z, n_trees, mtry, min_size, n_basis)
  wts1 <- weights(forest, matrix(1))
  wts2 <- weights(forest, matrix(2))

  expect_true(all(wts1[x != 1] == 0))
  expect_true(all(wts2[x != 2] == 0))
  expect_true(all((wts1 == 0) | (wts2 == 0)))
})

File: python/tests/test_edge_cases.py
import numpy as np
import rfcde
import pytest


def test_node_size_is_respected():
    n = 1000
    x = np.random.random((n, 1))
    z = np.random.random(n)

    n_trees = 1
    mtry = 1
    n_basis = 15
    for min_size in [1, 2, 3, 10, 100]:
        forest = rfcde.RFCDE(n_trees=n_trees,
                             mtry=mtry,
                             node_size=min_size,
                             n_basis=n_basis)
        forest.train(x, z)
        assert sum(forest.weights(x[0, :])) >= min_size
